sources/curriculum/0_before_you_start/workflows-and-pipelines-detailed/README.md

```diff
@@ -0,0 +1,194 @@
+
+1. **ETL**: extracting data from its source, transforming it, then loading it into a database.
+2. **Pre-process data**: This might include imputing missing values and choosing the training and testing sets.
+3. **Train the model(s)**: You can try different algorithms, features, and
+so on.
+4. Assess performance on the test set: Using an appropriate
+metric (e.g. $Precision@k$, $Recall$, $AUC$), examine the performance of your model "out of
+sample."
+5. Think of new things to try. Repeat steps 1 through 4 as appropriate.
+
+If the code base is not structured and named well, you might be struggling
+to remember the details of each step once you have built a few models. 
+What features did you use for each?
+What training and testing split?  What hyperparameters?
+
+Your code might be getting messy too. Did you overwrite the code for
+the previous model? Maybe you copied, pasted, and edited code from an
+earlier model. Can you still read what's there? It can quickly become
+a hodgepodge that requires heroics to decipher.
+
+In this session, we will introduce a workflow that can avoid (or at least reduce) these problems.
+
+# Data pipelines
+
+It is helpful to structure the data into multiple layers.
+
+##### Raw
+The data we receive from the partners and external sources is the raw data. 
+Raw data is immutable. Quoting from the popular workflow package [Data Science Cookiecutter](https://drivendata.github.io/cookiecutter-data-science/#data-is-immutable):
+
+Don't ever edit your raw data, especially not manually, and especially not in Excel. 
+Don't overwrite your raw data. 
+Don't save multiple versions of the raw data. 
+Treat the data (and its format) as immutable. 
+The code you write should move the raw data through a pipeline to your final analysis. 
+You shouldn't have to run all of the steps every time you want to make a new figure (see Analysis is a DAG), but anyone should be able to reproduce the final products with only the code in src and the data in data/raw.
+
+
+##### Intermediate
+If the raw data is messy, it is advisable to create an intermediate layer that consists of tidy copies of the raw data.
+Typical situations where this is useful are:
+- Data is received in multiple different file types 
+- Data fields are not typed (e.g. csv files, excel) or poorly typed (dates as strings, inconsistent date formats)
+- Column names are unclear, have spaces, special characters, or there are no column names 
+
+The transformations from raw to intermediate should be limited to fix the issues mentioned above. 
+We should not combine different data sets or create calculated fields.
+This is reserved for the next layer.
+
+Typical file formats for the intermediate layer are a data base (e.g. `postgres`) or `parquet` files.
+
+##### Processed
+To perform the modelling work, the input data needs to be combined and enriched, for example by creating features. 
+The data sets that are created in this process are stored in the processed layer.
```

Comment:
So layer is the equivalent to `schema`? How are we linking this two concepts?

Reply:
Yes, in a data base you can realise the layers through schemas.

Reply:
Given that we are being database centric (because partners limitations), we should be more specific on this.

Reply:
Why do you think we're data base centric? At Imperial we have 2 projects where all the data is images and videos - so not in data base. Happy to point out that in Data bases layers are realised through schemas though. I have made the change.

Reply:
You will be storing metadata about images, models, and videos in a database, right?

Reply:
Happy to discuss with both of you about this. We can set a call tomorrow if you like.